sdk/python/lib/pulumi/x/automation/stack.py

```diff
@@ -0,0 +1,532 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import json
+import grpc
+from concurrent import futures
+from enum import Enum
+from datetime import datetime
+from typing import List, Any, Mapping, MutableMapping, Optional
+
+from .cmd import CommandResult, _run_pulumi_cmd, OnOutput
+from .config import ConfigValue, ConfigMap, _SECRET_SENTINEL
+from .errors import StackAlreadyExistsError, CommandError
+from .server import LanguageServer
+from .workspace import Workspace, PulumiFn
+from ...runtime.settings import _GRPC_CHANNEL_OPTIONS
+from ...runtime.proto import language_pb2_grpc
+
+_DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%S.%fZ'
+
+
+class ExecKind(str, Enum):
+    LOCAL = "auto.local"
+    INLINE = "auto.inline"
+
+
+class StackInitMode(Enum):
+    CREATE = "create"
+    SELECT = "select"
+    CREATE_OR_SELECT = "create_or_select"
+
+
+class OutputValue:
+    value: Any
+    secret: bool
+
+    def __init__(self, value: Any, secret: bool):
+        self.value = value
+        self.secret = secret
+
+    def __repr__(self):
+        return _SECRET_SENTINEL if self.secret else repr(self.value)
+
+
+OutputMap = MutableMapping[str, OutputValue]
+
+OpMap = MutableMapping[str, int]
+
+
+class UpdateSummary:
+    # pre-update info
+    kind: str
+    start_time: datetime
+    message: str
+    environment: Mapping[str, str]
+    config: ConfigMap
+
+    # post-update info
+    result: str
+    end_time: datetime
+    version: Optional[int]
+    deployment: Optional[str]
+    resource_changes: Optional[OpMap]
+
+    def __init__(self,
+                 kind: str,
+                 start_time: datetime,
+                 message: str,
+                 environment: Mapping[str, str],
+                 config: Mapping[str, dict],
+                 result: str,
+                 end_time: datetime,
+                 version: Optional[int] = None,
+                 deployment: Optional[str] = None,
+                 resource_changes: Optional[OpMap] = None):
+        self.kind = kind
+        self.start_time = start_time
+        self.end_time = end_time
+        self.message = message
+        self.environment = environment
+        self.result = result
+        self.Deployment = deployment
+        self.resource_changes = resource_changes
+        self.version = version
+        self.config: ConfigMap = {}
+        for key in config:
+            self.config[key] = ConfigValue(**config[key])
+
+    def __repr__(self):
+        return f"UpdateSummary(result={self.result!r}, version={self.version!r}, start_time='{self.start_time!s} UTC', end_time='{self.end_time!s} UTC')"
```

Comment:
Now that the times are `datetime` should these args be shown as strings here?

Reply:
The `repr()` for datetime objects isn't particularly helpful... and even pycharm uses the str representation for debug. I'll go ahead and use it as it is represented elsewhere (without the quotes around it and the extra UTC - but in that case the timezone info is not obvious from the `repr()` but maybe that's not a big deal)

<img width="1928" alt="Screen Shot 2021-01-12 at 1 40 45 PM" src="https://user-images.githubusercontent.com/24326455/104377836-e9ae0900-54db-11eb-8136-808507f3f2fc.png">

Reply:
Shouldn't we show whatever values you'd have to pass to create the same `datetime` instance?

```
UpdateSummary(result='succeeded', version=32, start_time=datetime(<todo>), end_time=datetime(<todo>))
```

?

Reply:
Yeah, I hear what you're saying. I'll implement `__str__` functions for each of these too because that's really the experience I'm trying to improve here.